garage/logger/outputs.py

```diff
@@ -0,0 +1,121 @@
+"""Contains the output classes for the logger.
+
+Each class is sent logger data and handles it itself.
+"""
+import abc
+import csv
+import datetime
+import os
+import sys
+
+import dateutil.tz
+
+from garage.logger.singleton_tabular import TabularInput
+from garage.misc.console import colorize, mkdir_p
+
+
+class LogOutput(abc.ABC):
+    """Abstract class for Logger Outputs."""
+
+    @property
+    def types_accepted(self):
+        """The types that the logger may pass to this output."""
+        return ()
+
+    @abc.abstractmethod
+    def log_output(self, data, **kwargs):
```

Comment:
Perhaps there's a better name for this API? record? sink? 

@naeioi @krzentner

Reply:
Why not just log()?

Reply:
I don't have a preference. `log_output` looks good to me.

Reply:
@pelillian @naeioi @krzentner 

Let's look at the vocabulary for the API so that we can make it consistent/easy for novices to reason about.

* The logger has _inputs_ (sources) and _outputs_ (sinks)
* Between the _inputs_ and _outputs_ sits, the _logger_ which mediates flow between these two sides
* A user _logs_ data by calling `logger.log()` with an _input type_. Most _inputs_ are Python primitives like `string`, but there are also special cases (e.g. `tabular`, probably some stuff for `numpy.ndarrays` later)
* _outputs_ subscribe to _input types_ they _accept_ by declaring a property which returns a collection of Python _input types_ they can _accept_
* the _logger_ tell the _outputs_ when to _dump_ their data to their backing stores (e.g. flush or record)

At issue here: what happens in between _log_ and _dump_, when the _logger_ passes an _input_ to an _output_?
Suggestions:
* `record()` 
* `accept()`
* `put()`
* `post()`

I am slightly partial to `put()` or `record()`

Another question: _dump_ or _flush_? Generally `dump()` implies a one-time write-to-backing, whereas `flush()` implies an incremental update. `commit()` is another option, but might imply atomicity we can't actually live up to.

Reply:
I don't have a strong preference. This method won't be defined or called much compared to the number of `log` calls. `log_output` is a fairly obvious name, so I don't really see a strong reason to change it. `record()` would also be fine, `put()` seems not so great, since it's often associated with networking or queuing (neither of which is happening here). `post()` and `accept()` have the same issues, but to a larger degree. Methods named `flush()` typically don't take parameters, and are generally associated with buffering, which isn't really the behavior here.

If the motivating reason is to have a different verb to describe the output's operation than the logger itself, then `record()` seems fine to me. If we don't think that's a major concern (and it doesn't seem that major to me), then `log_output` seems fine.